synapse/storage/databases/main/search.py

```diff
@@ -760,20 +781,193 @@ def _to_postgres_options(options_dict: JsonDict) -> str:
     return "'%s'" % (",".join("%s=%s" % (k, v) for k, v in options_dict.items()),)
 
 
-def _parse_query(database_engine: BaseDatabaseEngine, search_term: str) -> str:
-    """Takes a plain unicode string from the user and converts it into a form
-    that can be passed to database.
-    We use this so that we can add prefix matching, which isn't something
-    that is supported by default.
+@dataclass
+class Phrase:
+    phrase: List[str]
+
+
+class SearchToken(enum.Enum):
+    Not = enum.auto()
+    Or = enum.auto()
+    And = enum.auto()
+    LParen = enum.auto()
+    RParen = enum.auto()
```

Comment:
> `AND`, `OR`, `-` (negation) set operators and parentheses
>
> If on pgsql 11+, pass the user inputer to `websearch_to_tsquery`

It seems that `websearch_to_tsquery` doesn't actually support parentheses in query text (they get removed); it also doesn't seem to recognize `AND` as an operator:

```sql
$ SELECT websearch_to_tsquery('simple', 'fox AND (nope OR brown)');
       websearch_to_tsquery       
----------------------------------
 'fox' & 'and' & 'nope' | 'brown'
```

The underlying `to_tsquery` does support them though.

Note that (ignoring the bogus `AND`), this isn't correct due to operator precedence: `A & (B | C) == A & B | A & C`, where we have `A & B | C`.

Reply:
I don't think the complexity in supporting parentheses is worth it? It feels quite niche for someone to need it and for them to know about it, but :shrug: